lib/topological_inventory/ansible_tower/operations/source.rb

```diff
@@ -10,58 +10,113 @@ class Source
 
         def initialize(params = {}, request_context = nil)
-          @params  = params
-          @context = request_context
-          @source_id = nil
+          self.params          = params
+          self.request_context = request_context
+          self.source_id       = params['source_id']
         end
 
         def availability_check
-          source_id = params["source_id"]
-          unless source_id
-            logger.error("Missing source_id for the availability_check request")
-            return
-          end
+          return if params_missing?
 
-          source = ::SourcesApiClient::Source.new
-          source.availability_status = connection_check(source_id)
+          status, error_message = connection_status
 
-          begin
-            api_client.update_source(source_id, source)
-          rescue ::SourcesApiClient::ApiError => e
-            logger.error("Failed to update Source id:#{source_id} - #{e.message}")
-          end
-          logger.info("Source#availability_check completed: Source #{source_id} is #{source.availability_status}")
+          update_source_and_endpoint(status, error_message)
+
+          logger.info("Source#availability_check completed: Source #{source_id} is #{status}")
         end
 
         private
 
-        def connection_check(source_id)
-          endpoints = api_client.list_source_endpoints(source_id)&.data || []
-          endpoint = endpoints.find(&:default)
-          return STATUS_UNAVAILABLE unless endpoint
+        def required_params
+          %w[source_id]
+        end
 
-          endpoint_authentications = api_client.list_endpoint_authentications(endpoint.id.to_s).data || []
-          return STATUS_UNAVAILABLE if endpoint_authentications.empty?
+        def params_missing?
+          is_missing = false
+          required_params.each do |attr|
+            if (is_missing = params[attr].blank?)
+              logger.error("Source#availability_check - Missing #{attr} for the availability_check request")
+              break
+            end
+          end
 
-          auth_id = endpoint_authentications.first.id
-          auth = Core::AuthenticationRetriever.new(auth_id, identity).process
-          return STATUS_UNAVAILABLE unless auth
+          is_missing
+        end
+
+        def connection_status
+          return [STATUS_UNAVAILABLE, ERROR_MESSAGES[:endpoint_not_found]] unless endpoint
+          return [STATUS_UNAVAILABLE, ERROR_MESSAGES[:authentication_not_found]] unless authentication
+
+          connection_check
+        end
 
+        def connection_check
           connection = ::TopologicalInventory::AnsibleTower::Connection.new
-          connection = connection.connect(endpoint.host, auth.username, auth.password)
+          connection = connection.connect(endpoint.host, authentication.username, authentication.password)
           connection.api.version
 
-          STATUS_AVAILABLE
+          [STATUS_AVAILABLE, nil]
```

Comment:
Small nit, but this can stay as just `STATUS_AVAILABLE`, had to try it myself to make sure but heres what I found:
```ruby
[ ⚛ ] ~ $ irb
Welcome to IRB. You are using ruby 2.6.6p146 (2020-03-31 revision 67876) [x86_64-linux]. Have fun ;)
>> status, err = [:status, :error]
>> status #=> :status
>> err #=> :error
>> status, err = :status
>> status #=> :status
>> err #=> nil
>> 

```

EDIT: 
Although that would be a bit harder to read. I'm ok with either, explicit might be better than implicit in this case.

Reply:
me too, anyway it's interesting, didn't know about that